scripts: replace pykwalify with jsonschema

kartben opened the pull request:
Converted `arch-schema.yml`, `board-schema.yml`, and `soc-schema.yml` from PyYAML format to JSON Schema format, including baking in some of the validation rules that were deferred to ad hoc Python right into the schemas (e.g. mutual exclusivity of certain fields).

Not necessarily so much faster as I'd hoped, but certainly not slower :)

```console
$ hyperfine --warmup 3 --command-name=before --prepare 'git switch main' 'west boards' --command-name=after --prepare 'git switch json_schemas' 'west boards'
Benchmark 1: before
  Time (mean ± σ):     542.1 ms ±  10.9 ms    [User: 417.0 ms, System: 118.8 ms]
  Range (min … max):   528.4 ms … 559.1 ms    10 runs
 
Benchmark 2: after
  Time (mean ± σ):     516.7 ms ±  18.6 ms    [User: 391.6 ms, System: 117.7 ms]
  Range (min … max):   501.3 ms … 567.0 ms    10 runs
 
Summary
  after ran
    1.05 ± 0.04 times faster than before
```

pdgendt commented on scripts/schemas/arch-schema.json:
```diff
@@ -0,0 +1,32 @@
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "$id": "https://zephyrproject.org/schemas/arch-schema.json",
```
Prefer to add the project in the id, as I'd like to do the same for `west`

```suggestion
  "$id": "https://zephyrproject.org/schemas/zephyr/arch-schema.json",
```

pdgendt commented on scripts/list_boards.py:
```diff
@@ -20,9 +21,11 @@
-    board_schema = yaml.load(f.read(), Loader=SafeLoader)
+    board_schema = json.load(f)
+
+board_validator = jsonschema.Draft202012Validator(board_schema)
```
The schema reference is part of the schema itself, prefer to try to load it from the file

```suggestion
board_validator = jsonschema.validators.validator_for(board_schema)
```

pdgendt reviewed (commented):
Nice work! I have a slight preference for providing the schemas as yaml files, because we can add comments there.

[`jsonschema` doesn't care](https://python-jsonschema.readthedocs.io/en/stable/faq/#can-jsonschema-be-used-to-validate-yaml-toml-etc) as it uses Python objects, and yaml is a superset of json.

pdgendt commented on scripts/schemas/arch-schema.yaml:
Now that it's yaml again, comments can be added, copyright and license?

kartben committed:
boards: snps: use correct revision scheme

revisons must be major.minor.patch so add the missing patch

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
scripts: replace pykwalify with jsonschema

Converted `arch-schema.yml`, `board-schema.yml`, and `soc-schema.yml`
from PyYAML format to JSON Schema format, including baking in some of
the validation rules that were deferred to ad hoc Python right into the
schemas (e.g. mutual exclusivity of certain fields).

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
drivers: i2c: i2c_dw: drop unnecessary include

Including soc.h is causing issues - drop as it's not needed
Fixes e.g. the below:
west build -p -b em_starterkit@2.2.0/emsk_em7d \
     tests/drivers/build_all/w1 -T drivers.w1.build

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented on scripts/list_boards.py:
```diff
@@ -11,21 +11,24 @@
-BOARD_SCHEMA_PATH = str(Path(__file__).parent / 'schemas' / 'board-schema.yml')
+BOARD_SCHEMA_PATH = str(Path(__file__).parent / 'schemas' / 'board-schema.yaml')
 with open(BOARD_SCHEMA_PATH) as f:
-    board_schema = yaml.load(f.read(), Loader=SafeLoader)
+    board_schema = yaml.safe_load(f)
```
mmmh for some reason I thought I had noticed it was being used automatically when available but looks like I was wrong

pdgendt commented on scripts/list_hardware.py:
```diff
@@ -9,26 +9,31 @@
+SOC_SCHEMA_PATH = str(Path(__file__).parent / 'schemas' / 'soc-schema.yaml')
 with open(SOC_SCHEMA_PATH) as f:
-    soc_schema = yaml.load(f.read(), Loader=SafeLoader)
+    soc_schema = yaml.safe_load(f)
```
Undo this change too.

kartben replied:
duh - thanks :)

pdgendt commented on scripts/list_hardware.py:
```diff
@@ -9,26 +9,31 @@
+ARCH_SCHEMA_PATH = str(Path(__file__).parent / 'schemas' / 'arch-schema.yaml')
 with open(ARCH_SCHEMA_PATH) as f:
-    arch_schema = yaml.load(f.read(), Loader=SafeLoader)
+    arch_schema = yaml.safe_load(f)
```
Same here.

pdgendt commented on scripts/list_hardware.py:
```diff
@@ -9,26 +9,31 @@
+SOC_SCHEMA_PATH = str(Path(__file__).parent / 'schemas' / 'soc-schema.yaml')
 with open(SOC_SCHEMA_PATH) as f:
-    soc_schema = yaml.load(f.read(), Loader=SafeLoader)
+    soc_schema = yaml.load(f, Loader=SafeLoader)
```
Does it work without reading first?

```suggestion
    soc_schema = yaml.load(f.read(), Loader=SafeLoader)
```

kartben replied:
:| Sorry...

kartben commented on scripts/schemas/board-schema.yaml:
```diff
@@ -0,0 +1,255 @@
+# When possible, constraints and validation rules should be modeled directly in this file.
+
+$schema: "https://json-schema.org/draft/2020-12/schema"
+$id: "https://zephyrproject.org/schemas/zephyr/board-schema.json"
```
@pdgendt oops, this was left unnoticed :) I am tempted to make the ID `https://zephyrproject.org/schemas/zephyr/board` -- thoughts?

pdgendt replied:
Yes, sounds good!

kartben commented:
@tejlmand can you please have a look? Thanks!

kartben commented on scripts/list_boards.py:
```diff
@@ -260,19 +254,6 @@ def load_v2_boards(board_name, board_yml, systems):
-                    sys.exit(f'ERROR: Malformed "board" section in file: {board_yml.as_posix()}\n'
-                             "Cannot find required key 'revisions'. Path: '/board/revision.'")
-
-            mutual_exclusive = {'socs', 'variants'}
-            if len(mutual_exclusive - board.keys()) < 1:
-                sys.exit(f'ERROR: Malformed "board" section in file: {board_yml.as_posix()}\n'
-                         f'{mutual_exclusive} are mutual exclusive at this level.')
```
addressed

kartben replied:
wait no let me double check, I think I still have some loopholes in there

tejlmand commented on scripts/schemas/board-schema.yaml:
```diff
@@ -0,0 +1,255 @@
+  runners:
+    $ref: "#/$defs/runnerSchema"
+# board and boards are mutually exclusive
+# TODO: is there really a reason to not allow having both or is it some legacy thing?
```
iirc the only reason for making them mutual exclusive was to keep parsing of the yaml file simpler.

Most important is that users gets an error if they specify `board` twice by mistake.

cfriedt commented:
@kartben - is this basically ready to go? Waiting for any other approvals or conversations to be resolved?
